cmd/outputs.go

```diff
@@ -0,0 +1,274 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2021 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/afero"
+
+	"github.com/loadimpact/k6/cloudapi"
+	"github.com/loadimpact/k6/lib"
+	"github.com/loadimpact/k6/lib/consts"
+	"github.com/loadimpact/k6/loader"
+	"github.com/loadimpact/k6/output"
+	"github.com/loadimpact/k6/output/extensions"
+	"github.com/loadimpact/k6/output/json"
+	"github.com/loadimpact/k6/stats"
+	"github.com/loadimpact/k6/stats/cloud"
+	"github.com/loadimpact/k6/stats/csv"
+	"github.com/loadimpact/k6/stats/datadog"
+	"github.com/loadimpact/k6/stats/influxdb"
+	"github.com/loadimpact/k6/stats/kafka"
+	"github.com/loadimpact/k6/stats/statsd"
+)
+
+// TODO: move this to an output sub-module after we get rid of the old collectors?
+//nolint: funlen
+func getAllOutputConstructors() (map[string]func(output.Params) (output.Output, error), error) {
+	// Start with the built-in outputs
+	result := map[string]func(output.Params) (output.Output, error){
+		"json": json.New,
```

Comment:
I started that way, but then I refactored it the way it currently is, for several different reasons, and I'd advocate maybe refactoring the JS extensions as well in the future. Here are the reasons:
1. This allows us to print much better error messages instead of panicking when there are name conflicts, see [below](https://github.com/loadimpact/k6/blob/7e1bc5dcab910f0d197a63e843e5591de4c60423/cmd/outputs.go#L129).
2. It allows us to very easily distinguish between built-in outputs and extension, which might be very useful if we want to send the latter in the usage reports.
3. Package `init()` functions are a bit of an antipattern (see [this](https://leighmcculloch.com/posts/tool-go-check-no-globals-no-inits/), for example), and in this case it's completely unnecessary to have them for built-in modules.
4. This doesn't look more elegant than a static map, to me it looks ugly as hell, I'd normally consider such imports code smells in a Go program :sweat_smile: :  https://github.com/loadimpact/k6/blob/51790fc8df37113ac20b586b28051e21672dab16/js/modules.go#L23-L34

You have a valid argument for easier extension->core merges, but I think this is a minor concern. Cleaning up the extension would take 1 minute of just removing the `init()` from it and adding the same constructor that was referenced in it to this map instead

Reply:
I agree, but those are arguments for where `Register()` is being called, which could just as well happen in a way that avoids these anti-patterns. Where do you envision `Register()` being called from for external outputs now?

My main issue is there being a distinction between internal and external outputs, but if you think there are benefits to that, I'm OK with it.

Reply:
> Where do you envision Register() being called from for external outputs now?

For extensions it would have to be in `init()` again, that unfortunately can't be easily changed due to the xk6 architecture. It's the simplest way to go about it, and is good enough for extensions. We just don't _need_ to do the same in k6, we have perfect knowledge of the built-in outputs, we can even write a simple `//go:generate` helper to generate the list for us, if the number of built-in outputs ever grows substantially.

Reply:
I am mostly for everything to be done the same way (in this case in `init()`) so we don't by accident break something and then for it to work for our "special" internal ones but not for the extensions. 

Introducing externally developed extensions in the base k6 by removing `init` is not as interesting IMO but it also ... kind of prevents us from introducing module/outputs by still letting them be outside the `k6` repo(and be able to be imported as external) and by just importing them and registering as "internal". Or more accurately while we can import them and register them as internal, the fact that they have `init` basically means that they will be registered as non 'internal' as well, which I don't like ;).

I think I brought this up back when xk6 was designed but there is nothing stopping us from requiring that `parh.to/extension/module/name` to have a function `NewModule() interface{}` and for xk6 to generate `additional_extensions.go` file in the root of the `k6` project that just calls the correct `Register` function based on something ... for example the name or the returned type. This even means that a single "extension" can both register a js module and an output if we go the "name" variant. 

The above suggestion can still be made and combined with #1802 and/or #1858 both of which will likely require some changes to the already written extensions. 

With the above we can completely remove the `init` from anywhere but `additional_extensions.go` 
* or just have it instead of adding additional files to add to change `main.go` for example 
* or for that matter just generate static maps in the corresponding files as in [nedomi](https://github.com/ironsmile/nedomi/tree/master/handler)

Reply:
> we don't by accident break something and then for it to work for our "special" internal ones but not for the extensions.

I doubt we'll ever touch the extension registration after we have it set up once, and it already uses the same interfaces as internal modules, so this is mostly an theoretical concern. In any case, the proper way to ensure extension compatibility would be to to have a CI task that actually uses k6. I created https://github.com/loadimpact/k6/issues/1873

> This even means that a single "extension" can both register a js module and an output if we go the "name" variant.

I am confused, this can easily be done now :confused: A single extension repo can register however many extension types and modules it wants to. Right now xk6 essentially does a `import _ whatever` and that's all it does, I don't think it cares one bit for what every extension does. So I don't see a benefit for the extra complexity.